lib/src/modules/status.cc

```diff
@@ -134,63 +143,102 @@ ActionOutcome Status::callAction(const ActionRequest& request) {
 
-    auto metadata_file = (results_dir_path / "metadata").string();
+    bool running_by_pid { false };
+    bool not_running_by_pid { false };
+    auto pid_file = (results_dir_path / "pid").string();
+    std::string pid_txt {};
+    int pid {};
+
+    // Read the PID file before the metadata, to avoid a race in the
+    // non-blocking task thread, due to the elapsed time between the
+    // end of the job and the metadata write
+    if (!fs::exists(pid_file)) {
+        LOG_DEBUG("PID file '%1%' does not exist", pid_file);
+    } else if (!lth_file::read(pid_file, pid_txt)) {
+        LOG_ERROR("Failed to read PID file '%1%'", pid_file);
+    } else if (pid_txt.empty()) {
+        LOG_ERROR("PID file '%1%' is empty", pid_file);
+    } else {
+        try {
+            pid = std::stoi(pid_txt);
+            // NOTE(ale): processExists() does not throw
+            if (Util::processExists(pid)) {
+                // NOTE(ale): will set status to 'running' iff
+                // metadata is valid and metadata.completed == false
+                running_by_pid = true;
+            } else {
+                not_running_by_pid = true;
+                bool cached { false };
+                {
+                    ResultsMutex::LockGuard a_l {
+                        ResultsMutex::Instance().access_mtx };  // RAII
+                    cached = ResultsMutex::Instance().exists(t_id);
+                }
+                if (cached) {
+                    // The process doe not exist anymore, but its
+                    // mutex is still cached - wait a bit before
+                    // reading the metadata to allow the non-blocking
+                    // to lock its mutex and update the metadata
+                    pcp_util::this_thread::sleep_for(
+                        pcp_util::chrono::milliseconds(METADATA_RACE_MS));
+                }
+            }
+        } catch (const std::invalid_argument& e) {
+            LOG_ERROR("Invalid value '%1%' stored in PID file '%2%'",
+                      pid_txt, pid_file);
+        }
+    }
+
     ActionMetadata metadata {};
+    auto metadata_file = (results_dir_path / "metadata").string();
 
     try {
-        metadata = ActionMetadata(metadata_file);
+        // Acquire the result mutex lock, if the transaction is cached
+        ResultsMutex::Mutex_Ptr mtx_ptr;
+        {
+            ResultsMutex::LockGuard a_l {
+                ResultsMutex::Instance().access_mtx };  // RAII
+            if (ResultsMutex::Instance().exists(t_id)) {
+                mtx_ptr = ResultsMutex::Instance().get(t_id);
+            }
+        }
+        if (mtx_ptr != nullptr) {
+            LOG_TRACE("Non-blocking transaction %1% mutex is cached; "
+                      "locking it", t_id);
+            ResultsMutex::LockGuard r_l { *mtx_ptr };  // RAII
+            metadata = ActionMetadata(metadata_file);
+        } else {
+            LOG_TRACE("Non-blocking transaction %1% mutex is not cached", t_id);
```

Comment:
This is safe because no child holds a mutex, and the current thread is the only one that would cause one to start and claim the mutex?